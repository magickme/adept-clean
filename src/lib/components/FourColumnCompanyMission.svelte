<script lang="ts">
    import logo from '$lib/assets/logos/magickme_title_cerulean.svg';
    import defaultFooterData from '$lib/collections/footer/links.json';
  
    const { dataTestid = '', footerData = defaultFooterData } = $props();
  
    const currentYear = new Date().getFullYear();
  
    const { socialLinks = [], footerColumns = [] } = footerData;

    interface SocialLink {
        name: string;
        href: string;
        icon: string;
        newTab?: boolean;
    }

    interface FooterItem {
        name: string;
        href: string;
        newTab?: boolean;
    }

    interface FooterColumn {
        title: string;
        items: FooterItem[];
    }
</script>
  
{#snippet socialIcon({ name, href, icon, newTab }: SocialLink)}
    <a {href} class="text-ivory hover:text-cerulean" target={newTab ? "_blank" : "_self"} rel={newTab ? "noopener noreferrer" : ""}>
        <span class="sr-only">{name}</span>
        <svg class="h-6 w-6" fill="currentColor" viewBox="0 0 24 24" aria-hidden="true">
            <path fill-rule="evenodd" d={icon} clip-rule="evenodd" />
        </svg>
    </a>
{/snippet}
  
{#snippet footerColumn({ title, items }: FooterColumn)}
    <div class={title !== 'Solutions' ? 'mt-10 md:mt-0' : ''}>
        <h3 class="text-h4 text-ivory font-montserrat">{title}</h3>
        <ul role="list" class="mt-6 space-y-4">
            {#each items as { name, href, newTab }}
                <li>
                    <a {href} class="text-cap text-ivory hover:text-cerulean font-sourceSans" target={newTab ? "_blank" : "_self"} rel={newTab ? "noopener noreferrer" : ""}>{name}</a>
                </li>
            {/each}
        </ul>
    </div>
{/snippet}
  
<footer class="bg-kohl" aria-labelledby="footer-heading" data-testid={dataTestid}>
    <h2 id="footer-heading" class="sr-only">Footer</h2>
    <div class="mx-auto max-w-7xl px-6 pb-8 pt-16 sm:pt-24 lg:px-8 lg:pt-32">
        <div class="xl:grid xl:grid-cols-3 xl:gap-8">
            <div class="space-y-8">
                <img 
                    class="h-7" 
                    src={logo} 
                    alt="Magick.Me Logo"
                    width="140"
                    height="28" 
                />
                <p class="text-sm leading-6 text-ivory font-sourceSans">Making the world a better place by democratizing access to spirituality.</p>
                <div class="flex space-x-6">
                    {#each socialLinks as link}
                        {@render socialIcon(link)}
                    {/each}
                </div>
            </div>
            <div class="mt-16 grid grid-cols-2 gap-8 xl:col-span-2 xl:mt-0">
                <div class="md:grid md:grid-cols-2 md:gap-8">
                    {@render footerColumn(footerColumns[0])}
                    {@render footerColumn(footerColumns[2])}
                </div>
            </div>
        </div>
        <div class="mt-16 border-t border-ivory/10 pt-8 sm:mt-20 lg:mt-24">
            <p class="text-cap text-ivory font-sourceSans">Â© {currentYear} Ultraculture Inc. All rights reserved.</p>
        </div>
    </div>
</footer>