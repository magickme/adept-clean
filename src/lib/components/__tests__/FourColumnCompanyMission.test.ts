import { describe, it, expect, vi, beforeEach } from 'vitest';
import { render, screen } from '@testing-library/svelte/svelte5';
import FourColumnCompanyMission from '../FourColumnCompanyMission.svelte';
import type { FooterData } from '$lib/types';

// Setup mocks before any data usage
vi.mock('$lib/assets/logos/magickme_title_cerulean.svg', () => ({
  default: '/mock/logo.svg'
}));

vi.mock('$lib/collections/footer/links.json', () => ({
  default: {
    socialLinks: [
      { name: 'Twitter', href: 'https://twitter.com', icon: 'path', newTab: true },
      { name: 'LinkedIn', href: 'https://linkedin.com', icon: 'path', newTab: true }
    ],
    footerColumns: [
      {
        title: 'Solutions',
        items: [
          { name: 'Link 1', href: '/link1', newTab: false },
          { name: 'Link 2', href: '/link2', newTab: true }
        ]
      },
      {
        title: 'Support',
        items: [
          { name: 'Help', href: '/help', newTab: false },
          { name: 'Contact', href: '/contact', newTab: false }
        ]
      }
    ]
  }
}));

// Mock data defined after mocks
const mockData: FooterData = {
  socialLinks: [
    { name: 'Twitter', href: 'https://twitter.com', icon: 'path', newTab: true },
    { name: 'LinkedIn', href: 'https://linkedin.com', icon: 'path', newTab: true }
  ],
  footerColumns: [
    {
      title: 'Solutions',
      items: [
        { name: 'Link 1', href: '/link1', newTab: false },
        { name: 'Link 2', href: '/link2', newTab: true }
      ]
    },
    {
      title: 'Support',
      items: [
        { name: 'Help', href: '/help', newTab: false },
        { name: 'Contact', href: '/contact', newTab: false }
      ]
    }
  ]
};

describe('FourColumnCompanyMission', () => {
  beforeEach(() => {
    vi.resetModules();
  });

  describe('Basic Rendering', () => {
    it('renders the component', () => {
      render(FourColumnCompanyMission, {
        props: {
          dataTestid: 'footer',
          footerData: mockData
        }
      });
      
      expect(screen.getByTestId('footer')).toBeTruthy();
    });

    it('renders the logo', () => {
      render(FourColumnCompanyMission, {
        props: { footerData: mockData }
      });
      
      const logo = screen.getByRole('img', { name: /Magick\.Me Logo/i });
      expect(logo).toBeTruthy();
      expect(logo.getAttribute('src')).toBe('/mock/logo.svg');
    });

    it('renders the mission statement', () => {
      render(FourColumnCompanyMission, {
        props: { footerData: mockData }
      });
      
      expect(screen.getByText(/Making the world a better place/)).toBeTruthy();
    });

    it('handles empty footer data', () => {
      render(FourColumnCompanyMission, {
        props: { 
          footerData: {
            socialLinks: [],
            footerColumns: []
          }
        }
      });
      
      expect(screen.getByRole('img', { name: /Magick\.Me Logo/i })).toBeTruthy();
      expect(screen.getByText(/Making the world a better place/)).toBeTruthy();
    });

    it('handles missing newTab property', () => {
      const incompleteData = {
        ...mockData,
        socialLinks: [{ name: 'Twitter', href: 'https://twitter.com', icon: 'path' }]
      };
      
      render(FourColumnCompanyMission, {
        props: { footerData: incompleteData }
      });
      
      const link = screen.getByRole('link', { name: 'Twitter' });
      expect(link.getAttribute('target')).toBe('_self');
    });

    it('matches snapshot', () => {
      const { container } = render(FourColumnCompanyMission, {
        props: { footerData: mockData }
      });
      
      expect(container).toMatchSnapshot();
    });
  });

  describe('Social Links', () => {
    it('renders all social links', () => {
      render(FourColumnCompanyMission, {
        props: { footerData: mockData }
      });

      mockData.socialLinks.forEach(link => {
        const socialLink = screen.getByRole('link', { name: link.name });
        expect(socialLink).toBeTruthy();
        expect(socialLink.getAttribute('href')).toBe(link.href);
        expect(socialLink.getAttribute('target')).toBe(link.newTab ? '_blank' : '_self');
      });
    });

    it('renders social icons', () => {
      render(FourColumnCompanyMission, {
        props: { footerData: mockData }
      });

      mockData.socialLinks.forEach(link => {
        const icon = screen.getByRole('link', { name: link.name }).querySelector('svg');
        expect(icon).toBeTruthy();
      });
    });
  });

  describe('Footer Columns', () => {
    it('renders all footer columns', () => {
      render(FourColumnCompanyMission, {
        props: { footerData: mockData }
      });

      mockData.footerColumns.forEach(column => {
        expect(screen.getByRole('heading', { name: column.title })).toBeTruthy();
        column.items.forEach(item => {
          const link = screen.getByRole('link', { name: item.name });
          expect(link).toBeTruthy();
          expect(link.getAttribute('href')).toBe(item.href);
        });
      });
    });
  });

  describe('Copyright', () => {
    it('renders current year', () => {
      render(FourColumnCompanyMission, {
        props: { footerData: mockData }
      });

      const year = new Date().getFullYear();
      expect(screen.getByText(new RegExp(year.toString()))).toBeTruthy();
    });
  });

  describe('Accessibility', () => {
    it('has correct ARIA attributes', () => {
      render(FourColumnCompanyMission, {
        props: { footerData: mockData }
      });
      
      expect(screen.getByRole('contentinfo')).toBeTruthy(); // footer element
      expect(screen.getByLabelText('Footer')).toBeTruthy();
    });

    it('has correct link attributes for external links', () => {
      render(FourColumnCompanyMission, {
        props: { footerData: mockData }
      });

      const externalLinks = screen.getAllByRole('link')
        .filter(link => link.getAttribute('target') === '_blank');
        
      externalLinks.forEach(link => {
        expect(link.getAttribute('rel')).toBe('noopener noreferrer');
      });
    });
  });
}); 