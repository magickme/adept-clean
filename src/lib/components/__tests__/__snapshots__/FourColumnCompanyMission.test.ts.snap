// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`FourColumnCompanyMission > Basic Rendering > matches snapshot 1`] = `
<div>
  <footer
    aria-labelledby="footer-heading"
    class="bg-kohl"
    data-testid=""
  >
    <h2
      class="sr-only"
      id="footer-heading"
    >
      Footer
    </h2>
     
    <div
      class="mx-auto max-w-7xl px-6 pb-8 pt-16 sm:pt-24 lg:px-8 lg:pt-32"
    >
      <div
        class="xl:grid xl:grid-cols-3 xl:gap-8"
      >
        <div
          class="space-y-8"
        >
          <img
            alt="Magick.Me Logo"
            class="h-7"
            src="/mock/logo.svg"
          />
           
          <p
            class="text-sm leading-6 text-ivory"
          >
            Making the world a better place by democratizing access to spirituality.
          </p>
           
          <div
            class="flex space-x-6"
          >
            <a
              class="text-ivory hover:text-cerulean"
              href="https://twitter.com"
              rel="noopener noreferrer"
              target="_blank"
            >
              <span
                class="sr-only"
              >
                Twitter
              </span>
               
              <svg
                aria-hidden="true"
                class="h-6 w-6"
                fill="currentColor"
                viewBox="0 0 24 24"
              >
                <path
                  clip-rule="evenodd"
                  d="path"
                  fill-rule="evenodd"
                />
              </svg>
            </a>
            <a
              class="text-ivory hover:text-cerulean"
              href="https://linkedin.com"
              rel="noopener noreferrer"
              target="_blank"
            >
              <span
                class="sr-only"
              >
                LinkedIn
              </span>
               
              <svg
                aria-hidden="true"
                class="h-6 w-6"
                fill="currentColor"
                viewBox="0 0 24 24"
              >
                <path
                  clip-rule="evenodd"
                  d="path"
                  fill-rule="evenodd"
                />
              </svg>
            </a>
            
          </div>
        </div>
         
        <div
          class="mt-16 grid grid-cols-2 gap-8 xl:col-span-2 xl:mt-0"
        >
          <div
            class="md:grid md:grid-cols-2 md:gap-8"
          >
            <div>
              <h3
                class="text-h4 text-ivory"
              >
                Solutions
              </h3>
               
              <ul
                class="mt-6 space-y-4"
                role="list"
              >
                <li>
                  <a
                    class="text-cap text-ivory hover:text-cerulean"
                    href="/link1"
                    rel=""
                    target="_self"
                  >
                    Link 1
                  </a>
                </li>
                <li>
                  <a
                    class="text-cap text-ivory hover:text-cerulean"
                    href="/link2"
                    rel="noopener noreferrer"
                    target="_blank"
                  >
                    Link 2
                  </a>
                </li>
                
              </ul>
            </div>
            <div
              class="mt-10 md:mt-0"
            >
              <h3
                class="text-h4 text-ivory"
              >
                Support
              </h3>
               
              <ul
                class="mt-6 space-y-4"
                role="list"
              >
                <li>
                  <a
                    class="text-cap text-ivory hover:text-cerulean"
                    href="/help"
                    rel=""
                    target="_self"
                  >
                    Help
                  </a>
                </li>
                <li>
                  <a
                    class="text-cap text-ivory hover:text-cerulean"
                    href="/contact"
                    rel=""
                    target="_self"
                  >
                    Contact
                  </a>
                </li>
                
              </ul>
            </div>
            
          </div>
           
          <div
            class="md:grid md:grid-cols-2 md:gap-8"
          >
            
          </div>
        </div>
      </div>
       
      <div
        class="mt-16 border-t border-ivory/10 pt-8 sm:mt-20 lg:mt-24"
      >
        <p
          class="text-cap text-ivory"
        >
          Â© 2024 Ultraculture Inc. All rights reserved.
        </p>
      </div>
    </div>
  </footer>
  
</div>
`;
